### PR TITLE
Fix region control flow to functional pass with incompatible cast.

### LINKED ISSUE #46656: Issue with tflite converter in tf-nightly:  'tf.If' op 'then_branch' input type tensor<?xf32> is incompatible with input type tensor<?xi64> at index 0
**System information**
- OS Platform and Distribution: Linux Ubuntu 18.04.5
- Python version: 3.7.4
- TensorFlow version: tf-nightly-2.5.0.dev20210124

Hello, I have a new issue with the tflite converter which is happening only when using nightly (it isn't happening when using tensorflow main branch). I tried as much as I can to create the minimal code to reproduce: 

    !pip install tf-nightly
    import tensorflow as tf
    import os
    import numpy as np
    
    class EXAMPLE(tf.Module):
    
        @tf.function(input_signature=[tf.TensorSpec(shape=[100], dtype=tf.float32)])
        def calculate(self, x):
        
          maxima_ind = tf.where(x > 0.8)
          maxima_ind = tf.gather(maxima_ind,0,axis=1)
    
          maxima_ind = tf.cast(maxima_ind, dtype=tf.float32)
    
          if len(maxima_ind) > 10:
              maxima_ind = tf.cast(maxima_ind, dtype=tf.float32)
    
          return maxima_ind
          
    to_export = EXAMPLE()
    np.random.seed(54)
    buffer_size = 100
    x1  = tf.convert_to_tensor(np.random.rand(buffer_size).astype('float32'))
    
    
    solution = to_export.calculate(x1)
    print(solution)
    models_dir = '/content/model_example/'
    tf.saved_model.save(to_export, models_dir)
    imported = tf.saved_model.load(models_dir)
    converter = tf.lite.TFLiteConverter.from_saved_model(models_dir) # path to the SavedModel directory
    converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS,
                                            tf.lite.OpsSet.SELECT_TF_OPS]
    
    tflite_model = converter.convert()
    with open(models_dir + 'model_example.tflite', 'wb') as f:
      f.write(tflite_model)

When running it, I'm getting this error: 

    Exception                                 Traceback (most recent call last)
    /usr/local/lib/python3.6/dist-packages/tensorflow/lite/python/convert.py in toco_convert_protos(model_flags_str, toco_flags_str, input_data_str, debug_info_str, enable_mlir_converter)
        216                                                  debug_info_str,
    --> 217                                                  enable_mlir_converter)
        218       return model_str
    
    4 frames
    Exception: <unknown>:0: error: loc("cond@__inference_calculate_3155"): 'tf.If' op 'then_branch' input type tensor<?xf32> is incompatible with input type tensor<?xi64> at index 0
    
    
    During handling of the above exception, another exception occurred:
    
        ConverterError                            Traceback (most recent call last)
        /usr/local/lib/python3.6/dist-packages/tensorflow/lite/python/convert.py in toco_convert_protos(model_flags_str, toco_flags_str, input_data_str, debug_info_str, enable_mlir_converter)
            218       return model_str
            219     except Exception as e:
        --> 220       raise ConverterError(str(e))
            221 
            222   if distutils.spawn.find_executable(_toco_from_proto_bin) is None:
        
        ConverterError: <unknown>:0: error: loc("cond@__inference_calculate_3155"): 'tf.If' op 'then_branch' input type tensor<?xf32> is incompatible with input type tensor<?xi64> at index 0

You can also use this gist (which include the above code): https://colab.research.google.com/gist/AiaHaruv/6cac62614ef85f76db8bcfcfa0b8baae/conversion_bug.ipynb
I understand there is a thing with the types but I'm using casting before the if statement and inside of it I'm actually doing nothing, and it isn't happening in the main branch..  
Any help will be appreciated. Thank you

### PR DESCRIPTION
**Pull request by WindQAQ**
Fixes https://github.com/tensorflow/tensorflow/issues/46656.

When matching call args, do not skip incompatible cast. Otherwise, there will be incompatible args passed to build control flow op.

Also note that the https://github.com/tensorflow/tensorflow/issues/46656 is broken after https://github.com/tensorflow/tensorflow/commit/f9818f12c332e8a35dbc9042af55649be64f30c4 (so it works in 2.4.x). Before f9818f12c332e8a35dbc9042af55649be64f30c4, the single call op with few operations is inlined so it won't fall into

https://github.com/tensorflow/tensorflow/blob/e323dc9a05193c76ba5db1c71d0ef019be676224/tensorflow/compiler/mlir/tensorflow/transforms/region_control_flow_to_functional.cc#L287-L291

### CODE REVIEW

File: tensorflow/compiler/mlir/tensorflow/transforms/region_control_flow_to_functional.cc [hunk 1]
```diff
@@ -217,8 +217,18 @@ bool MatchCallArgs(CallOp first, CallOp second, ArgMatcherFn matcher) {
...
     // Get the defining Op, skipping over casts.
     auto get_defining_op = [](Value value) {
-      while (llvm::isa_and_nonnull<CastOp>(value.getDefiningOp()))
-        value = cast<CastOp>(value.getDefiningOp()).getOperand();
+      while (auto cast_op =
+                 llvm::dyn_cast_or_null<CastOp>(value.getDefiningOp())) {
+        // Consider cast compatibility in case
+        //    %cast = "tf.Cast"(%0) : (tensor<2xi64>) -> tensor<2xf32>
+        // is skipped.
+        if (AreCastCompatible(
```
Comment (smit-hinsu) [inline tensorflow/compiler/mlir/tensorflow/transforms/region_control_flow_to_functional.cc:225]: AreCastCompatible is used for shape verification. I think you want to verify that source element type of operand and result are the same.

**Review by smit-hinsu — CHANGES_REQUESTED** [review]
Thanks for the investigating and proposing the fix!

Solutions looks good to me. Adding Rahul who is more familiar with this.

File: tensorflow/compiler/mlir/tensorflow/transforms/region_control_flow_to_functional.cc (hunk 1 above)
Comment (WindQAQ) [reply tensorflow/compiler/mlir/tensorflow/transforms/region_control_flow_to_functional.cc:225]: I think it checks both element type and shape and is used https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc#L792 to verfiy IfOp/WhileOp.

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/tensorflow/ir/tf_types.cc#L251-L268

Comment (smit-hinsu) [reply tensorflow/compiler/mlir/tensorflow/transforms/region_control_flow_to_functional.cc:225]: We can expect all legal tensorflow tf.Cast ops to have the same operand and result shape at runtime. We don't have a verifier for this but we should add one.

File: tensorflow/compiler/mlir/tensorflow/transforms/region_control_flow_to_functional.cc [hunk 2]
```diff
@@ -217,8 +217,18 @@ bool MatchCallArgs(CallOp first, CallOp second, ArgMatcherFn matcher) {
...
+                llvm::ArrayRef<Type>{cast_op.SrcT(), cast_op.DstT()})) {
```
Comment (jurahul) [inline tensorflow/compiler/mlir/tensorflow/transforms/region_control_flow_to_functional.cc:226]: Should we also check cast_op.Truncate()?

File: tensorflow/compiler/mlir/tensorflow/tests/region-control-flow-to-functional.mlir [hunk 3]
```diff
@@ -168,6 +168,50 @@ func @testIf2Result(%arg0: tensor<i1>, %arg1: tensor<2xf32>) -> tensor<2xf32> {
 
 // -----
 
+// Do not skip incompatible cast for trivial transform.
```
Comment (jurahul) [inline tensorflow/compiler/mlir/tensorflow/tests/region-control-flow-to-functional.mlir:171]: Not sure if this test is actually testing the changed code. We look through casts inside the IfRegion body, but the cast here is outside.

**Review by jurahul — COMMENTED** [review]
Thanks for fixing this! Please see comments below.

File: tensorflow/compiler/mlir/tensorflow/transforms/region_control_flow_to_functional.cc (hunk 1 above)
Comment (WindQAQ) [reply tensorflow/compiler/mlir/tensorflow/transforms/region_control_flow_to_functional.cc:225]: Got it!

File: tensorflow/compiler/mlir/tensorflow/tests/region-control-flow-to-functional.mlir (hunk 3 above)
Comment (WindQAQ) [reply tensorflow/compiler/mlir/tensorflow/tests/region-control-flow-to-functional.mlir:171]: Thanks for review! This testcase is exactly the one in the original issue. MatchCallArgs does not limit to ops inside the region, that is, it keeps tracking back the definingOp until it is not a cast op.

Comment (WindQAQ) [reply tensorflow/compiler/mlir/tensorflow/tests/region-control-flow-to-functional.mlir:171]: Updated comment: Two testcases for IfOp are trivial transforms but one with extern incompatible cast and the other has non-extern incompatible cast.

File: tensorflow/compiler/mlir/tensorflow/transforms/region_control_flow_to_functional.cc [hunk 4]
```diff
@@ -217,8 +217,16 @@ bool MatchCallArgs(CallOp first, CallOp second, ArgMatcherFn matcher) {
...
     // Get the defining Op, skipping over casts.
     auto get_defining_op = [](Value value) {
-      while (llvm::isa_and_nonnull<CastOp>(value.getDefiningOp()))
-        value = cast<CastOp>(value.getDefiningOp()).getOperand();
+      while (auto cast_op =
+                 llvm::dyn_cast_or_null<CastOp>(value.getDefiningOp())) {
+        // Consider cast compatibility in case
+        //    %cast = "tf.Cast"(%0) : (tensor<2xi64>) -> tensor<2xf32>
+        // is skipped.
+        if ((cast_op.SrcT() != cast_op.DstT()) || cast_op.Truncate()) {
```
Comment (jurahul) [inline tensorflow/compiler/mlir/tensorflow/transforms/region_control_flow_to_functional.cc:225]: if src and dest type are same, then I'd assume Truncate = false, so no need to check that.

Comment (WindQAQ) [reply tensorflow/compiler/mlir/tensorflow/transforms/region_control_flow_to_functional.cc]: Updated!

**Review by jurahul — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Pull request merged**